be/src/olap/rowset/segment_v2/column_reader.cpp

```diff
@@ -466,6 +486,42 @@ Status ArrayFileColumnIterator::next_batch(size_t* n, ColumnBlockView* dst, bool
     return Status::OK();
 }
 
+Status ArrayFileColumnIterator::_seek_by_offsets(ordinal_t ord) {
+    // using offsets info
+    ordinal_t offset = 0;
+    RETURN_IF_ERROR(_peek_one_offset(&offset));
+    RETURN_IF_ERROR(_item_iterator->seek_to_ordinal(offset));
+    return Status::OK();
+}
+
+Status ArrayFileColumnIterator::seek_to_ordinal(ordinal_t ord) {
+    RETURN_IF_ERROR(_offset_iterator->seek_to_ordinal(ord));
+    if (_array_reader->is_nullable()) {
+        RETURN_IF_ERROR(_null_iterator->seek_to_ordinal(ord));
+    }
+    return _seek_by_offsets(ord);
+}
+
+Status ArrayFileColumnIterator::_caculate_offsets(ssize_t start,
+                                                  vectorized::MutableColumnPtr& offsets,
+                                                  size_t* num_items) {
+    auto& column_offsets = static_cast<vectorized::ColumnArray::ColumnOffsets&>(*offsets);
+    ordinal_t last_offset = 0;
+    RETURN_IF_ERROR(_peek_one_offset(&last_offset));
+
+    auto& offsets_data = column_offsets.get_data();
+    *num_items = last_offset - offsets_data[start];
+    // caculate real offsets
+    ordinal_t first_offset = offsets_data[start - 1]; // -1 is valid
+    ordinal_t first_ord = offsets_data[start];
+    for (ssize_t i = start; i + 1 < offsets_data.size(); ++i) {
+        offsets_data[i] = first_offset + (offsets_data[i + 1] - first_ord);
+    }
```

Comment:
Why should we re-calculate the offsets here? Some questions here:

1. Does `vectorized::MutableColumnPtr& dst` in `ArrayFileColumnIterator::next_batch` share some data with others?
2. Does the `offsets_data` read from different pages cause the re-calculation?

Reply:
Ok, that is the scenario `ArrayFileColumnIterator::read_by_rowids`. What about the scenario in read data sequently?

Reply:
1. 2 both answer is yes

Reply:
Maybe there is a potential optimization point if we don't read by rowids. It seems that the item_writer doesn't reset the next_rowid in the same segment. So I don't think these answer are yes.

Reply:
sequenctly also should be recaculate, eg. (0, 1, 3, ....100, 103, 109, 114) , assume the offset=100 row_id = 20, we only need row_id from [20, 23), so after recaculate the offset became( 0, 3, 9)

Reply:
Ok.